flips/20220203-capability-controllers.md

```diff
@@ -0,0 +1,539 @@
+
+If a path that was unlinked is linked again all the capabilities created from that path resume working. 
+This can be used to redirect a capability to a different object, 
+but can also be dangerous if done unintentionally, 
+reviving a previously revoked capability.
+
+### Capabilities are values
+
+Capabilities are a value type. This means that any capability that an account has access to can be copied, 
+and potentially given to someone else. 
+The copied capability will use the link on the same path as the original capability.
+
+## Problem statement
+
+The current capability API has a few pain points: 
+
+- The concepts capabilities/linking/revoking/redirecting are hard to understand for new developers that are coming to Flow, 
+as that is a lot of new concepts to grasp before one can get started.
+
+- Issuing and managing private capabilities that can be revoked at a granular level, 
+  by creating a custom private linked path for each capability, is difficult.
+
+- Unless the path name includes some hints, there is no way to know when a path was created, 
+  thus making it more difficult to remember why a certain capability was issued.
+
+- If an old path (that has been unlinked) is accidentally reused and re-linked, 
+  this will revive a capability that is supposed to be revoked.
+
+## Suggested change
+
+The suggested change addresses these pain points by:
+
+- Removing (or abstracting away) the concept of links.
+- Making it easier to create new capabilities (with individual links).
+- Offering a way to iterate through capabilities on a storage path.
+- Removing the need to have a `/private` domain.
+- Introducing Capability Controllers (CapCons) that handle management of capabilities.
+
+### Capability Controllers (CapCons)
+
+Each Capability would have an associated CapCon that would be responsible for managing the Capability. 
+The CapCon would be created when the Capability is issued (created). 
+If the Capability is copied (since it is a value type) it shares the CapCon with the original 
+(i.e., calling revoke on the CapCon revokes all copies of the associated capability).
+
+A CapCon and all the copies of the Capability that CapCon controls would have the same ID.
+Capability/CapCon ids are unique per account.   
+
+The data associated with CapCons would be stored in arrays, 
+so that each storage path on an account has an array of CapCons of Capabilities issued from that storage path. 
+CapCons are a non-storable object (similar to AuthAccounts).
+
+The definition of the `CapabilityController`.
+
+```cadence
+struct CapabilityController {
+    /// The block height when the capability was created.
+    let issueHeight: UInt64
+   
+    /// The Type of the capability, i.e.: the T in Capability<T>.
+    let borrowType: Type
+   
+    /// The id of the related capability.
+    /// This is the UUID of the created capability.
+    /// All copies of the same capability will have the same UUID
+    let capabilityID: UInt64
```

Comment:
One thing to note is that UUIDs are 128-bit, not 64 as mentioned here. Capability ID's are account-local, so making them 64-bit shouldn't be an issue (i fear the person with 1.8e19 capabilities). Thoughts?

Reply:
Good point! The "UUID" naming in comment is actually wrong, as pointed out by @dete a few times before for resource "UUID"s: they are not unique, but rather local to the chain. 

I'll update the comment

Reply:
Fixed in [d3bfc4a](https://github.com/onflow/flow/pull/798/commits/d3bfc4a4de6e981c573b71983690797e87e63087)